alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/healthprocessor/plugins/solr/MutableHealthReport.java

```diff
@@ -34,43 +37,46 @@ public MutableHealthReport(@NonNull NodeHealthStatus healthStatus, @Nonnull Node
-        this.healthStatus = healthStatus;
-        if (forced) {
-            forcedHealthStatus = healthStatus;
-        }
-        messages.add(reason);
+    public void addHealthReport(EndpointHealthStatus healthStatus, NodeRef.Status nodeRefStatus,
+            SearchEndpoint searchEndpoint) {
+        addHealthReport(new EndpointHealthReport(healthStatus, nodeRefStatus, searchEndpoint));
     }
 
-    public void markUnhealthy(String reason) {
-        mark(NodeHealthStatus.UNHEALTHY, reason, true);
+    public void addHealthReport(EndpointHealthReport endpointHealthReport) {
+        if (healthStatus != null) {
+            throw new IllegalStateException(
+                    "Can not add endpoint health reports when a health status is set directly.");
+        }
+        endpointHealthReports.add(endpointHealthReport);
     }
 
-    public void markUnknownForced(String reason) {
-        mark(NodeHealthStatus.NONE, reason, true);
-    }
+    private NodeHealthStatus getHealthStatus() {
+        if (healthStatus != null) {
+            return healthStatus;
+        }
 
-    public void markUnknown(String reason) {
-        if (healthStatus == null) {
-            mark(NodeHealthStatus.NONE, reason, false);
-        } else {
-            mark(healthStatus, reason, false);
+        EndpointHealthStatus highestHealthStatus = EndpointHealthStatus.UNSET;
+        for (EndpointHealthReport endpointHealthReport : endpointHealthReports) {
+            EndpointHealthStatus healthStatus = endpointHealthReport.getHealthStatus();
+            if (healthStatus.ordinal() < highestHealthStatus.ordinal()) {
+                highestHealthStatus = healthStatus;
+            }
         }
-    }
 
-    public void markHealthy(String reason) {
-        if (healthStatus == null || healthStatus == NodeHealthStatus.NONE) {
-            mark(NodeHealthStatus.HEALTHY, reason, false);
-        } else {
-            mark(healthStatus, reason, false);
+        if (highestHealthStatus == EndpointHealthStatus.UNSET) {
+            throw new IllegalStateException("Can not have no health status set and have no endpoint health reports");
         }
+        return highestHealthStatus.getNodeHealthStatus();
     }
 
     public NodeHealthReport getHealthReport() {
-        Objects.requireNonNull(healthStatus, "healthStatus");
-        if (forcedHealthStatus != null) {
-            return new NodeHealthReport(forcedHealthStatus, nodeRef, messages);
-        }
-        return new NodeHealthReport(healthStatus, nodeRef, messages);
+        Set<String> allMessages = Stream.concat(
+                messages.stream(),
+                endpointHealthReports.stream().map(EndpointHealthReport::getMessage)
+        ).collect(Collectors.toSet());
+
+        NodeHealthReport nodeHealthReport = new NodeHealthReport(getHealthStatus(), nodeRef, allMessages);
+        nodeHealthReport.data(EndpointHealthReport.class).addAll(endpointHealthReports);
```

Comment:
I can call it `createHealthReport()` instead if you think that is a better name?

The reason why I only create a NodeHealthReport here is because it is **immutable**, and we can not change health status or messages after the class is created.
Immutability is important to limit the places where a health report can be modified, as they are sent to all fixers & reporters in sequence. It would be quite a problem if these components would be able to modify the health report directly.